cpp/src/arrow/python/arrow_to_python.cc

```diff
@@ -47,6 +47,11 @@ Status DeserializeList(PyObject* context, std::shared_ptr<Array> array, int64_t
                        const std::vector<std::shared_ptr<Tensor>>& tensors,
                        PyObject** out);
 
+Status DeserializeSet(PyObject* context, std::shared_ptr<Array> array, int64_t start_idx,
```

Comment:
As a matter of hygiene, can you change these functions to pass `array` as `const Array& array`? You may need to change the cast inside the macro. It may also improve performance